runtime/runtime/src/deterministic_account_id.rs

```diff
@@ -0,0 +1,182 @@
+}
+
+impl AccountState {
+    fn of(maybe_account: Option<&Account>) -> AccountState {
+        let Some(account) = maybe_account else {
+            return AccountState::NonExist;
+        };
+
+        if account.contract().is_none() {
+            return AccountState::Uninit;
+        }
+
+        AccountState::Active
+    }
+}
+
+pub(crate) fn action_deterministic_state_init(
+    state_update: &mut TrieUpdate,
+    apply_state: &ApplyState,
+    account: &mut Option<Account>,
+    account_id: &AccountId,
+    receipt: &Receipt,
+    action: &DeterministicStateInitAction,
+    result: &mut ActionResult,
+) -> Result<(), RuntimeError> {
+    let current_protocol_version = apply_state.current_protocol_version;
+    match AccountState::of(account.as_ref()) {
+        AccountState::NonExist => {
+            let storage_usage_config = &apply_state.config.fees.storage_usage_config;
+            create_deterministic_account(account, storage_usage_config);
+            deploy_deterministic_account(
+                state_update,
+                account.as_mut().expect("account must exists now"),
+                account_id,
+                &action.state_init,
+                result,
+                current_protocol_version,
+            )?;
+        }
+        AccountState::Uninit => {
+            deploy_deterministic_account(
+                state_update,
+                account.as_mut().expect("account must exists in uninit"),
+                account_id,
+                &action.state_init,
+                result,
+                current_protocol_version,
+            )?;
+        }
+        AccountState::Active => {
+            // Account already exists, do nothing.
+        }
+    }
+    debug_assert!(
+        matches!(AccountState::of(account.as_ref()), AccountState::Active),
+        "account must be active now"
+    );
+
+    // Use attached deposit to satisfy storage staking requirements and refund
+    // the rest.
+    let account = account.as_mut().expect("account must exist now");
+    let deposit_refund = match check_storage_stake(account, account.amount(), &apply_state.config) {
+        Ok(_) => {
+            // no additional storage needed, refunding all
+            action.deposit
+        }
+        Err(StorageStakingError::LackBalanceForStorageStaking(missing_amount)) => {
+            if missing_amount <= action.deposit {
+                // use exactly as much as needed and refund the rest
+                let new_balance = safe_add_balance(account.amount(), missing_amount)?;
+                account.set_amount(new_balance);
+                action
+                    .deposit
+                    .checked_sub(missing_amount)
+                    .expect("just checked missing_amount <= action.deposit")
+            } else {
+                // not enough balance at this point -> no refund
+                // (following actions might be able to fix the balance  requirements but these will no affect this refund)
```

Comment:
typo and double space: `(following actions might be able to fix the balance requirements but these will not affect this refund)`